### PR TITLE
Add Adafruit Macropad RP2040

### PR DESCRIPTION
**Pull request by Jpe230**
## Description

Add support for the Adafruit Macropad RP2040.

This PR includes a driver for the SPI OLED that uses a modified version of the core driver with the difference of the transport protocol, users can use this driver the same as the core driver.

Also this board uses Adafruit proprietary VID but the folks over Adafruit were kind to generate a new pair for exclusive use of this board in QMK: 
image

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by PeterFalken** [comment]
It's great to see the Adafruit Macropad RP2040 get supported in QMK.

File: keyboards/adafruit/macropad/config.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+/* Double tap the side button to enter bootloader */
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET_LED GP13
+#define RP2040_BOOTLOADER_DOUBLE_TAP_RESET_TIMEOUT 500U
+
+/* Audio (Unsupported for now)*/
+// #define AUDIO_PIN GP16
+// #define SPEAKER_SHUTDOWN GP14
+
+#include "config_led.h"
```
Comment (zvecr) [inline keyboards/adafruit/macropad/config.h:85]:
```suggestion
```

File: keyboards/adafruit/macropad/config_led.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#ifdef RGB_MATRIX_ENABLE
```
Comment (zvecr) [inline keyboards/adafruit/macropad/config_led.h:19]: should be moved to `config.h`

File: keyboards/adafruit/macropad/macropad.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "quantum.h"
+
+#define KCN KC_NO
```
Comment (zvecr) [inline keyboards/adafruit/macropad/macropad.h:21]:
```suggestion
#define ___ KC_NO
```

File: keyboards/adafruit/macropad/macropad.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+// clang-format off
+ #define LAYOUT(  \
+             K02, \
+   K10, K11, K12, \
+   K20, K21, K22, \
+   K30, K31, K32, \
+   K40, K41, K42  \
+ ) \
+ { \
+   { KCN, KCN, K02 }, \
```
Comment (zvecr) [inline keyboards/adafruit/macropad/macropad.h:32]:
```suggestion
   { ___, ___, K02 }, \
```

**Review by zvecr — CHANGES_REQUESTED** [review]
Can we change the `oled_spi` folder to `lib` so its consistent with other keyboard structures.

File: keyboards/adafruit/macropad/config_led.c [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+#include "macropad.h"
+#include "rgb_matrix.h"
+#include "config_led.h"
+
+#define NA NO_LED
+
+/* RGB Positioning */
+led_config_t g_led_config = { {
+             { NA }, /* Encoder Button */
```
Comment (zvecr) [inline keyboards/adafruit/macropad/config_led.c:27]:
```suggestion
    { NA, NA, NA }, /* Encoder Button */
```

File: keyboards/adafruit/macropad/config_led.c [hunk 6]
```diff
@@ -0,0 +1,43 @@
+/* Copyright 2022 Jose Pablo Ramirez <jp.ramangulo@gmail.com>
```
Comment (zvecr) [inline keyboards/adafruit/macropad/config_led.c:1]: should renamed to `macropad.c`

**Comment by Jpe230** [comment]
Solved and rebased :D

File: keyboards/adafruit/macropad/config_led.c [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifdef RGB_MATRIX_ENABLE
+
+#include "macropad.h"
+#include "rgb_matrix.h"
```
Comment (zvecr) [inline keyboards/adafruit/macropad/config_led.c:20]:
```suggestion

#include "macropad.h"

#ifdef RGB_MATRIX_ENABLE
```

File: keyboards/adafruit/macropad/rules.mk [hunk 8]
```diff
@@ -0,0 +1,27 @@
+# Project specific files
+SRC += \
+	config_led.c \
+	lib/ssd1306_sh1106.c
+
+
```
Comment (zvecr) [inline keyboards/adafruit/macropad/rules.mk:6]:
```suggestion
SRC += lib/ssd1306_sh1106.c
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/adafruit/macropad/info.json [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+                {"label":"BackSpace", "x":2, "y":3},
+                {"label":"KC_7", "x":0, "y":2},
+                {"label":"KC_8", "x":1, "y":2},
+                {"label":"KC_9", "x":2, "y":2},
+                {"label":"KC_4", "x":0, "y":1},
+                {"label":"KC_5", "x":1, "y":1},
+                {"label":"KC_6", "x":2, "y":1},
+                {"label":"KC_1", "x":0, "y":0},
+                {"label":"KC_2", "x":1, "y":0},
+                {"label":"KC_3", "x":2, "y":0}
```
Comment (zvecr) [inline keyboards/adafruit/macropad/info.json:20]:
```suggestion
                {"label":"Mute", "x":2, "y":0},
                {"label":"Enter", "x":0, "y":1},
                {"label":"KC_0", "x":1, "y":1},
                {"label":"BackSpace", "x":2, "y":1},
                {"label":"KC_7", "x":0, "y":2},
                {"label":"KC_8", "x":1, "y":2},
                {"label":"KC_9", "x":2, "y":2},
                {"label":"KC_4", "x":0, "y":3},
                {"label":"KC_5", "x":1, "y":3},
                {"label":"KC_6", "x":2, "y":3},
                {"label":"KC_1", "x":0, "y":4},
                {"label":"KC_2", "x":1, "y":4},
                {"label":"KC_3", "x":2, "y":4}
```

**Comment by Jpe230** [comment]
VScode had a fart and autoclosed the PR :( Sorry for that

**Comment by tzarc** [comment]
For future reference (mainly for others, here) -- so long as there's no proliferation and use of the SPI OLED driver elsewhere, we're okay with it.

Quantum Painter is the preferred display abstraction for ARM devices going forward, and whilst it's unfortunate this SPI OLED isn't yet compatible I do ask that once it is (it's in my pipeline), this gets board swapped over to QP.

**Comment by Jpe230** [comment]
> For future reference (mainly for others, here) -- so long as there's no proliferation and use of the SPI OLED driver elsewhere, we're okay with it.
> 
> 
> 
> Quantum Painter is the preferred display abstraction for ARM devices going forward, and whilst it's unfortunate this SPI OLED isn't yet compatible I do ask that once it is (it's in my pipeline), this gets board swapped over to QP.
> 
> 
> 
> EDIT: for clarity -- keymaps within this board are fine; we want to compartmentalise the OLED driver over time.

Thanks for the input Tzarc! Will do once support for this display in QP drops!

File: keyboards/adafruit/macropad/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+#include <stdio.h>
```
Comment (drashna) [inline keyboards/adafruit/macropad/keymaps/default/keymap.c:18]:
```suggestion
```

File: keyboards/adafruit/macropad/macropad.c [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    { 0, 42 }, { 112, 42 }, { 224, 42},
+    { 0, 64 }, { 112, 64 }, { 224, 64}
+}, {
+    4, 4, 4,
+    4, 4, 4,
+    4, 4, 4,
+    4, 4, 4
+} };
+
+#endif
```
Comment (drashna) [inline keyboards/adafruit/macropad/macropad.c:42]: This allows for a default behavior to be defined, without the need for the user to specify the behavior.  This is very useful for the configurator which doesn't have a way to configure this behavior, currently.

```suggestion
#endif

#ifdef ENCODER_ENABLE
bool encoder_update_kb(uint8_t index, bool clockwise) {
    if (!encoder_update_user(index, clockwise)) { return false; }
    if (index == 0) {
        if (clockwise) {
            tap_code_delay(KC_VOLU, 10);
        } else {
            tap_code_delay(KC_VOLD, 10);
        }
    }
    return true;
}
#endif
```

File: keyboards/adafruit/macropad/readme.md [hunk 12]
```diff
@@ -0,0 +1,29 @@
+# Adafruit MacroPad RP2040
+
+![AdafruitMacropad](https://i.imgur.com/dSBSwcJ.jpeg)
```
Comment (drashna) [inline keyboards/adafruit/macropad/readme.md:3]:
```suggestion
![AdafruitMacropad](https://i.imgur.com/dSBSwcJh.jpeg)
```

File: keyboards/adafruit/macropad/readme.md [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+A RP2040-powered Macropad with a 3x4 layout.
+
+- Keyboard Maintainer: [Jpe230](https://github.com/jpe230/) 
+- Hardware Supported: Adafruit MacroPad RP2040
+- Hardware Availability: [Barebones kit](https://www.adafruit.com/product/5100) [Starter Kit](https://www.adafruit.com/product/5128)
+
+Make example for this board (after setting up your build environment):
```
Comment (drashna) [inline keyboards/adafruit/macropad/readme.md:11]: Could you add instructions on how to enter the bootloader, to flash new firmware? 

Eg: 
```markup
## Bootloader

Enter the bootloader in 3 ways:

* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
* **Physical reset button**: Briefly press the button on the back of the PCB - some may have pads you must short instead
* **Keycode in layout**: Press the key mapped to `RESET` if it is available
```

File: keyboards/adafruit/macropad/rules.mk [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+NKRO_ENABLE        = yes
+WS2812_DRIVER      = vendor
+ENCODER_ENABLE     = yes
+RGB_MATRIX_ENABLE  = yes
+RGB_MATRIX_DRIVER  = WS2812
+EEPROM_DRIVER      = transient
+OLED_ENABLE        = yes
+OLED_DRIVER        = custom
+EXTRAKEY_ENABLE    = yes
+# AUDIO_DRIVER       = pwm_software
```
Comment (drashna) [inline keyboards/adafruit/macropad/rules.mk:25]:
```suggestion
# MCU name
MCU = RP2040

# Bootloader selection
BOOTLOADER = rp2040

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
NKRO_ENABLE = yes           # Enable N-Key Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
# AUDIO_DRIVER = pwm_software

ENCODER_ENABLE = yes
RGB_MATRIX_ENABLE = yes
RGB_MATRIX_DRIVER = WS2812
WS2812_DRIVER = vendor
EEPROM_DRIVER = transient
OLED_ENABLE = yes
OLED_DRIVER = custom

# Project specific files
SRC += lib/ssd1306_sh1106.c

QUANTUM_LIB_SRC += spi_master.c
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/adafruit/macropad/readme.md (hunk 12 above)
Comment (Jpe230) [reply keyboards/adafruit/macropad/readme.md:3]: Pardon my ignorance but why the "h"?

Comment (drashna) [reply keyboards/adafruit/macropad/readme.md]: No worries.  It's a neat little hack for imgur.  It renders a lower resolution version that is better for the readme pages. 

Though, double checking, probably not needed here.

Comment (Jpe230) [reply keyboards/adafruit/macropad/readme.md]: Whoopsie 😅 should I revert it? Or leave your suggestion?

**Comment by Jpe230** [comment]
Rebased with lastest changes from develop (mainly for EEPROM Emulation) - this required the removal of the transient driver, can confirm from my end that it works wonderfully!

image

**Review by drashna — APPROVED** [review]

**Comment by jepler** [comment]
I tested this and it's working great! Thank you!

**Review by zvecr — APPROVED** [review]

**Comment by Jpe230** [comment]
Thanks for the merge! ❤️

**Pull request merged**